test/extensions/filters/http/cache/http_cache_implementation_test_common.cc

```diff
@@ -0,0 +1,747 @@
+  LookupRequest request = makeLookupRequest(request_path);
+  LookupContextPtr context = cache().makeLookupContext(std::move(request), decoder_callbacks_);
+  absl::Notification lookup_result_received;
+  context->getHeaders([this, &lookup_result_received](LookupResult&& result) {
+    lookup_result_ = std::move(result);
+    lookup_result_received.Notify();
+  });
+  EXPECT_TRUE(lookup_result_received.WaitForNotificationWithTimeout(absl::Seconds(3)));
+  return context;
+}
+
+absl::Status HttpCacheImplementationTest::insert(LookupContextPtr lookup,
+                                                 const Http::TestResponseHeaderMapImpl& headers,
+                                                 const absl::string_view body,
+                                                 std::chrono::milliseconds timeout) {
+  return insert(std::move(lookup), headers, body, absl::nullopt, timeout);
+}
+
+absl::Status HttpCacheImplementationTest::insert(
+    LookupContextPtr lookup, const Http::TestResponseHeaderMapImpl& headers,
+    const absl::string_view body, const absl::optional<Http::TestResponseTrailerMapImpl> trailers,
+    std::chrono::milliseconds timeout) {
+  InsertContextPtr inserter = cache().makeInsertContext(std::move(lookup), encoder_callbacks_);
+  const ResponseMetadata metadata{time_source_.systemTime()};
+  bool headers_end_stream = body.empty() && !trailers.has_value();
+  inserter->insertHeaders(headers, metadata, headers_end_stream);
+
+  if (headers_end_stream) {
+    return absl::OkStatus();
+  }
+  if (!body.empty() && !trailers.has_value()) {
+    inserter->insertBody(Buffer::OwnedImpl(body), nullptr,
+                         /*end_stream=*/!trailers.has_value());
+    return absl::OkStatus();
+  }
+  if (body.empty() && trailers.has_value()) {
+    inserter->insertTrailers(trailers.value());
+    return absl::OkStatus();
+  }
+
+  // For responses with body and trailers, wait for insertBody's callback before
+  // calling insertTrailers.
+  std::promise<absl::Status> insert_promise;
+  auto insert_mutex = std::make_shared<absl::Mutex>();
+  auto insert_cancelled = std::make_shared<absl::Notification>();
+  InsertCallback insert_callback = [&trailers, &inserter, &insert_promise, insert_mutex,
+                                    insert_cancelled](bool success_ready_for_more) {
+    absl::MutexLock lock(insert_mutex.get());
+    if (insert_cancelled->HasBeenNotified()) {
+      return;
+    }
+    if (!success_ready_for_more) {
+      insert_promise.set_value(absl::UnknownError("Insert was aborted by cache"));
+    }
+    inserter->insertTrailers(trailers.value());
+    insert_promise.set_value(absl::OkStatus());
+  };
+  inserter->insertBody(Buffer::OwnedImpl(body), insert_callback,
+                       /*end_stream=*/!trailers.has_value());
+
+  std::future<absl::Status> insert_future = insert_promise.get_future();
+  if (insert_future.wait_for(timeout) != std::future_status::ready) {
```

Comment:
I think this won't work with simulated time. See test/test_common/test_time_system.h and use TestTimeSystem::waitFor instead.

I'm hoping this will go smoothly!

Reply:
This is here to let the test fail gracefully if the cache never calls the insert-done callback, instead of waiting forever. That means it does have to be wall-clock time, since there's no guarantee that advancing the simulated clock will make the callback fire.

Reply:
Tests pass fine with TestTimeSystem::waitFor, but I'm getting a format script error that I can't find a way around:

```
ERROR: ./test/extensions/filters/http/cache/http_cache_implementation_test_common.cc:138: Don't reference real-world time sources; use TimeSystem::advanceTime(Wait|Async)
```

Do you know what's wrong there?

Reply:
Yes -- the lint script is expecting to see time_system_.waitFor here. I think if it's not too much of a hassle you could do a replacement of `time_source_` for `time_system_` in the test infrastructure. You could do that in a separate trivial rename PR first, or just roll it into this one.

I think the reason it was labeled `time_source_` is that the prototype caching infrastructure was co-developed with what we now call TimeSystem, but at the time, the Timer semantics was rolled into the TimeSource API, before we decided that TimeSource, which is used in production code, should only have the methods to query the current system/monotonic time, and we should separate out the abstraction for messing with timers into TimeSystem, and for doing waits into TestTimeSystem. 

Here's the hacky Python code that you are running into:
```
    # We want to look for a call to condvar.waitFor, but there's no strong pattern
    # to the variable name of the condvar. If we just look for ".waitFor" we'll also
    # pick up time_system_.waitFor(...), and we don't want to return true for that
    # pattern. But in that case there is a strong pattern of using time_system in
    # various spellings as the variable name.
    def has_cond_var_wait_for(self, line):
        wait_for = line.find(".waitFor(")
        if wait_for == -1:
            return False
        preceding = line[0:wait_for]
        if preceding.endswith("time_system") or preceding.endswith("timeSystem()") or \
          preceding.endswith("time_system_"):
            return False
        return True
```
Rather than making this hackier by also look for time_source and variants, I think we should just rename the variable-names in the caching infrastructure to be consistent with the rest of tests.

Reply:
Got it, renamed the variable. The error message for this case isn't great, so I'll send out a separate PR to fix that.